include/picongpu/plugins/saxs/Saxs.kernel

```diff
@@ -0,0 +1,203 @@
+    struct KernelSaxs
+    {
+        /** Kernel function.
+         *
+         * @param acc alpaka accelerator
+         * @param cellsGrid Dimensions of BORDER + CORE in cells, not super
+         *      cells.
+         * @param densityBoxGPU Data box of the density device storage, shifted
+         *      to exclude 1st GUARD.
+         * @param globalOffset Offset from the global to the local domain.
+         * @param fieldPos TmpField in cell position.
+         * @param amplitudeBox Device side data box of the output Buffer.
+         * @param totalNumVectors Number of scattering vectors to process.
+         * @param scatteringVectors Scattering vectors to process.
+         * @param probingBeam Probing beam characterization.
+         * @param currentStep Current simulation step.
+         */
+        template<
+            typename T_Acc,
+            typename T_DensityBoxGPU,
+            typename T_FieldPos,
+            typename T_DBox,
+            typename T_TotalNumVectors,
+            typename T_ScatteringVectors,
+            typename T_ProbingBeam
+        >
+        DINLINE void operator( )(
+            T_Acc const & acc,
+            DataSpace< simDim > cellsGrid,
+            T_DensityBoxGPU densityBoxGPU,
+            DataSpace< simDim > globalOffset,
+            T_FieldPos fieldPos,
+            T_DBox amplitudeBox,
+            T_TotalNumVectors const totalNumVectors,
+            T_ScatteringVectors scatteringVectors,
+            T_ProbingBeam probingBeam,
+            uint32_t currentStep
+
+        ) const
+        {
+            constexpr uint32_t blockSize =
+                pmacc::math::CT::volume< SuperCellSize >::type::value;
+            constexpr uint32_t numWorkers = T_numWorkers;
+            uint32_t const workerIdx = threadIdx.x;
+
+            using complex_X = pmacc::math::Complex< float_X >;
+            using namespace pmacc::mappings::threads;
+            using namespace pmacc;
+            // Storage for positions in the beam coordinate system.
+            PMACC_SMEM(
+                acc,
+                positions,
+                memory::Array< float2_X, blockSize >
+            );
+            // Storage for (form factor density * beam intensity factor).
+            PMACC_SMEM(
+                acc,
+                densities,
+                memory::Array< float_X, blockSize >
+            );
+
+            uint32_t const linAccessBlockBegin =  blockIdx.x * blockSize;
+
+            ForEachIdx<
+                IdxConfig<
+                    blockSize,
+                    numWorkers
+                >
+            >{ workerIdx }(
+                [ & ](
+                    uint32_t const linearIdx,
+                    uint32_t const
+                )
+                {
+                    // Each thread reads one field value and saves it together
+                    // with its position, in UNIT_LENGTH, in the global domain.
+                    floatD_X fieldGlobalPosition;
```

Comment:
(Not important) this could be moved a couple of lines down where the actual values is assigned.